examples/step-82/step-82.cc

```diff
@@ -0,0 +1,1500 @@
+
+    std::cout << "Done. " << std::endl;
+  }
+
+
+
+  // @sect4{BiLaplacianLDGLift::assemble_matrix}
+
+  // This function assembles the matrix $A$ whose entries are defined
+  // by $A_{ij}=A_h(\varphi_j,\varphi_i)$ which involves the product of
+  // discrete Hessians and the penalty terms.
+  template <int dim>
+  void BiLaplacianLDGLift<dim>::assemble_matrix()
+  {
+    matrix = 0;
+
+    QGauss<dim>     quad(fe.degree + 1);
+    QGauss<dim - 1> quad_face(fe.degree + 1);
+
+    const unsigned int n_q_points      = quad.size();
+    const unsigned int n_q_points_face = quad_face.size();
+
+    FEValues<dim> fe_values(fe, quad, update_hessians | update_JxW_values);
+
+    FEFaceValues<dim> fe_face(
+      fe, quad_face, update_values | update_gradients | update_normal_vectors);
+
+    FEFaceValues<dim> fe_face_neighbor(
+      fe, quad_face, update_values | update_gradients | update_normal_vectors);
+
+    const unsigned int n_dofs = fe_values.dofs_per_cell;
+
+    std::vector<types::global_dof_index> local_dof_indices(n_dofs);
+    std::vector<types::global_dof_index> local_dof_indices_neighbor(n_dofs);
+    std::vector<types::global_dof_index> local_dof_indices_neighbor_2(n_dofs);
+
+    // As indicated in the introduction, the following matrices are used for
+    // the contributions of the products of the discrete Hessians.
+    FullMatrix<double> stiffness_matrix_cc(n_dofs,
+                                           n_dofs); // interactions cell / cell
+    FullMatrix<double> stiffness_matrix_cn(
+      n_dofs, n_dofs); // interactions cell / neighbor
+    FullMatrix<double> stiffness_matrix_nc(
+      n_dofs, n_dofs); // interactions neighbor / cell
+    FullMatrix<double> stiffness_matrix_nn(
+      n_dofs, n_dofs); // interactions neighbor / neighbor
+    FullMatrix<double> stiffness_matrix_n1n2(
+      n_dofs, n_dofs); // interactions neighbor1 / neighbor2
+    FullMatrix<double> stiffness_matrix_n2n1(
+      n_dofs, n_dofs); // interactions neighbor2 / neighbor1
+
+    // The following matrices are used for the contributions of the two
+    // penalty terms:
+    FullMatrix<double> ip_matrix_cc(n_dofs, n_dofs); // interactions cell / cell
+    FullMatrix<double> ip_matrix_cn(n_dofs,
+                                    n_dofs); // interactions cell / neighbor
+    FullMatrix<double> ip_matrix_nc(n_dofs,
+                                    n_dofs); // interactions neighbor / cell
+    FullMatrix<double> ip_matrix_nn(n_dofs,
+                                    n_dofs); // interactions neighbor / neighbor
+
+    std::vector<std::vector<Tensor<2, dim>>> discrete_hessians(
+      n_dofs, std::vector<Tensor<2, dim>>(n_q_points));
+    std::vector<std::vector<std::vector<Tensor<2, dim>>>>
+      discrete_hessians_neigh(GeometryInfo<dim>::faces_per_cell,
+                              discrete_hessians);
+
+    Tensor<2, dim> H_i, H_j;
+    Tensor<2, dim> H_i_neigh, H_j_neigh;
+    Tensor<2, dim> H_i_neigh2, H_j_neigh2;
```

Comment:
I don't think we need to declare these here.

Reply:
Do you mean that we should directly use `Tensor <2, dim> H_i = discrete_hessians[i][q];` at line 556 and similarly for the other occurrences?

Reply:
Yes. The current version is more `C`-style. Assigning a value at a single place allows you to make the variable const, which significantly improves the readability of codes since the origin of values is clear. I know that it is annoying to type out the type, but you could also use `const auto H_i = discrete_hessians[i][q];`